### ISSUE TITLE
Very Old Segments of a Deleted Table Still Present in Servers

### ISSUE THREAD

**Issue by suddendust** [apache/pinot#8182]
I had deleted `someTableName` long time back but was surprised to see its segments inside the data directory on servers.  One such segment that I see is`someTableName__0__248__20210212T1915Z`, which is almost a year old. We have also configured the retention period to be 1 day, so not really sure why this segment is still hanging around.

**Comment by Jackie-Jiang** [comment]
@walterddr Can you please take a look at this issue? We might need to change the order of segment deletion to ensure the segment from the deep store is deleted before removing the segment metadata, or the segment will become an orphan.

**Comment by walterddr** [comment]
yes. I think Jackie's analysis is probably what actually happened. could you please try querying with the following 
```
SELECT $segmentName, count(*) FROM <yourTable> WHERE $segmentName = '<yourTable>__0__248__20210212T1915Z' GROUP BY $segmentName
```
and see if it actually returns anything. I felt like the deep store data has already been clear and it is just the zkmetadata for the segment is still there dangling around.

**Comment by suddendust** [comment]
@walterddr This query isn't working as the table has been deleted. I get a 

```
[
  {
    "message": "TableDoesNotExistError",
    "errorCode": 190
  }
]
```

**Comment by walterddr** [comment]
yes this sounds like the deletion order issue. relates to #8092

**Comment by walterddr** [comment]
@suddendust, I am not working on it right now. 

the issue can be tricky to fix and some redesign of the deletion logic might be needed. (details I've listed in #8092). please kindly take a look and we can discuss what would be the best approach. 

Thanks,
Rong

**Comment by suddendust** [comment]
@walterddr Haven't had a chance to dive-deep into this, but one thing is confirmed that it's not just the metadata hanging around, it's complete segments. For example, the following view in this dashboard is backed by a Pinot table with retention of 1 day. I loaded this on 1st March 20:19 UTC. So this means I should see no data before 28th Feb 20:19 UTC. However, I see data for much before:

<img width="1426" alt="Screenshot 2022-03-02 at 1 49 09 AM" src="https://user-images.githubusercontent.com/84911643/156242785-707e2112-d14f-405f-bbfc-58b2383adbe9.png">

Confirmed that these segments also exist in the servers and ZK. 

Thanks!
